### PR TITLE
Enable Resolve Attr Path for List or Dict of Promise

### PR DESCRIPTION
**Pull request by Mecoli1219**
## Tracking issue
https://github.com/flyteorg/flyte/issues/5856

## Why are the changes needed?
We should support this:
```
@dataclass
class DC:
    a: int

@dataclass
class DC2:
    a: DC
    b: int

@task(container_image=image)
def task_1(in1: int) -> typing.List[int]:
    return list(range(max(2, in1)))

@task(container_image=image)
def task_2(in1: int, in2: typing.List[int]) -> int:
    return in1 + sum(in2)

@task(container_image=image)
def task_3() -> DC2:
    return DC2(a=DC(a=1), b=2)

@workflow
def my_wf(in1: int, in2: int) -> int:
    a = task_1(in1=in1)
    a1 = a[0]
    dc = task_3()
    b = task_2(in1=5, in2=[a1, dc["b"], dc.a["a"], in2])
    return b
```

## What changes were proposed in this pull request?
If the input is the list, call `resolve_attr_path_in_promise` for each element of that.

## How was this patch tested?
1. Run the code above locally & remotely.
2. A new unit test is added

### Screenshots
image

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

**Comment by Future-Outlier** [comment]
> cc @Future-Outlier 

Yes you are amazing

File: flytekit/core/promise.py [hunk 1]
```diff
@@ -99,6 +99,10 @@ def my_wf(in1: int, in2: int) -> int:
         try:
             if type(v) is Promise:
                 v = await resolve_attr_path_in_promise(v)
+            elif type(v) is list:
```
Comment (pingsutw) [inline flytekit/core/promise.py:102]: Should we use recursive here? What if we have a nested list.

File: tests/flytekit/unit/core/test_promise.py [hunk 2]
```diff
@@ -147,6 +148,45 @@ def t1(a: typing.Union[float, MyDataclass, Annotated[typing.List[typing.Any], Ba
...
+
+    i0 = "foo"
+    i1 = Promise("n1", TypeEngine.to_literal(ctx, "bar", str, TypeEngine.to_literal_type(str)))
+
+    @task
+    def t1(a: List[str]):
+        print(a)
+
+
+    lits = translate_inputs_to_literals(ctx, {"a": [
```
Comment (pingsutw) [inline tests/flytekit/unit/core/test_promise.py:182]: Can we add a test for the nested list?

File: flytekit/core/promise.py (hunk 1 above)
Comment (Mecoli1219) [reply flytekit/core/promise.py:102]: You are correct

File: tests/flytekit/unit/core/test_promise.py (hunk 2 above)
Comment (Mecoli1219) [reply tests/flytekit/unit/core/test_promise.py:182]: yes, done

**Review by pingsutw — DISMISSED** [review]

**Comment by Mecoli1219** [comment]
@pingsutw I updated the function to remove some redundant code and also support the Dictionary. Could you review it for me? Thx!

**Review by pingsutw — APPROVED** [review]

**Comment by Future-Outlier** [comment]
testing this PR, let me merge it.

**Comment by Future-Outlier** [comment]
It's great.
image

Do you want to also add a test for dict[str, type]?
@Mecoli1219

**Comment by Mecoli1219** [comment]
> Do you want to also add a test for dict[str, type]?

@Future-Outlier I updated it in the description

**Pull request merged**